File: thread_ring.exs
defmodule ThreadRing do
  def loop(next_pid) do
    receive do
      0 ->
        # Added LOG_END before halting
        IO.puts("LOG_END:#{System.system_time(:millisecond)}")
        IO.puts("--- FINISHED ---")
        System.halt(0)

      token ->
        send(next_pid, token - 1)
        loop(next_pid)
    end
  end

  def run do
    n_actors = 5_000
    token_value = 100_000_000

    IO.puts("--- STARTING THREAD RING (Elixir) ---")
    IO.puts("LOG_START:#{System.system_time(:millisecond)}")

    first_pid = Enum.reduce(n_actors..2, self(), fn _id, neighbor_pid ->
      spawn(fn -> loop(neighbor_pid) end)
    end)

    send(first_pid, token_value)

    loop(first_pid)
  end
end

ThreadRing.run()

File: ThreadRing.java
import akka.actor.AbstractActor;
import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import akka.actor.Props;
import java.util.ArrayList;

public class ThreadRing {

    static class Worker extends AbstractActor {
        private ActorRef next;

        @Override
        public Receive createReceive() {
            return receiveBuilder()
                .match(ActorRef.class, actorRef -> {
                    this.next = actorRef;
                })
                .match(Long.class, n -> {
                    if (n > 0) {
                        next.tell(n - 1L, getSelf());
                    } else {
                        // Added LOG_END before terminating
                        System.out.println("LOG_END:" + System.currentTimeMillis());
                        System.out.println("--- FINISHED ---");
                        getContext().getSystem().terminate();
                    }
                })
                .build();
        }
    }

    public static void main(String[] args) {
        int nActors = 1_000; 
        Long nPasses = 50_000_000L; 

        ActorSystem system = ActorSystem.create("ThreadRingSystem");
        
        System.out.println("--- STARTING THREAD RING ---");
        System.out.println("LOG_START:" + System.currentTimeMillis());

        ArrayList<ActorRef> ring = new ArrayList<>();
        for (int i = 0; i < nActors; i++) {
            ring.add(system.actorOf(Props.create(Worker.class)));
        }

        for (int i = 0; i < nActors; i++) {
            ActorRef current = ring.get(i);
            ActorRef next = ring.get((i + 1) % nActors);
            current.tell(next, ActorRef.noSender());
        }

        ring.get(0).tell(nPasses, ActorRef.noSender());
    }
}
